network/weightedroundrobin.go

```diff
@@ -58,23 +52,66 @@ func (r *WeightedRoundRobin) NextProxy(_ IConnWrapper) (IProxy, *gerr.GatewayDEr
 		return nil, gerr.ErrNoProxiesAvailable.Wrap(errors.New("proxy list is empty"))
 	}
 
-	var selected *weightedProxy
+	var selected IProxy
+	var maxWeight int
+	totalWeight := 0
+
+	// Prepare batch updates
+	updates := make(map[string]raft.WeightedProxy)
+
+	// Calculate weights and find the selected proxy
+	for _, proxy := range r.proxies {
+		proxyName := proxy.GetBlockName()
+		weight, exists := r.raftNode.Fsm.GetWeightedRRState(r.groupName, proxyName)
+		if !exists {
+			weight = raft.WeightedProxy{
+				EffectiveWeight: r.initialWeights[proxyName],
+				CurrentWeight:   0,
+			}
+		}
+
+		totalWeight += weight.EffectiveWeight
+		newWeight := weight.CurrentWeight + weight.EffectiveWeight
 
-	// Adjust weights and select the proxy with the highest current weight.
-	for _, p := range r.proxies {
-		p.currentWeight += p.effectiveWeight
-		if selected == nil || p.currentWeight > selected.currentWeight {
-			selected = p
+		// Store the update
+		updates[proxyName] = raft.WeightedProxy{
+			CurrentWeight:   newWeight,
+			EffectiveWeight: weight.EffectiveWeight,
 		}
+
+		if selected == nil || newWeight > maxWeight {
+			selected = proxy
+			maxWeight = newWeight
+		}
```

Comment:
Should we break the loop when we choose the proxy?

Reply:
- We need to sum up all effective weights to calculate totalWeight. This total is essential because it's used to decrease the selected proxy's current weight after selection (selectedProxy.CurrentWeight -= totalWeight).

- We can't stop at the first proxy that has a higher weight than the current maxWeight because a later proxy in the list might have an even higher weight. We must examine all proxies to ensure we find the one with the truly highest current weight.